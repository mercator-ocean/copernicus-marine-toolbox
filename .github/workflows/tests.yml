name: Run tests

on:
  pull_request:
    branches: [ "main", "copernicusmarine-toolbox-v2" ]

jobs:
  tests:
    runs-on: self-hosted
    timeout-minutes: 30

    steps:
      - name: Check out code
        uses: actions/checkout@v4

      - uses: mamba-org/setup-micromamba@v1
        with:
          micromamba-version: '1.5.6-0'
          micromamba-binary-path: ${{ runner.temp }}/bin/micromamba
          environment-file: conda_environment_test.yaml
          environment-name: copernicusmarine_test
          condarc-file: .condarc
          cache-environment: true
          post-cleanup: 'all'

      - name: Run tests
        env:
          COPERNICUSMARINE_SERVICE_USERNAME: ${{ secrets.COPERNICUSMARINE_SERVICE_USERNAME }}
          COPERNICUSMARINE_SERVICE_PASSWORD: ${{ secrets.COPERNICUSMARINE_SERVICE_PASSWORD }}
        run: make run-tests
        shell: micromamba-shell {0}
