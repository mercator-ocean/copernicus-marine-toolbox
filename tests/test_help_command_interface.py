from tests.test_utils import execute_in_terminal


class TestHelpCommandLineInterface:
    def test_help_from_describe_is_as_expected(self, snapshot):
        help_output = execute_in_terminal(
            ["copernicusmarine", "describe", "--help"]
        )

        assert str(help_output).split("\\n") == snapshot

    def test_help_from_get_is_as_expected(self, snapshot):
        help_output = execute_in_terminal(
            ["copernicusmarine", "get", "--help"]
        )

        assert str(help_output).split("\\n") == snapshot

    def test_help_from_subset_is_as_expected(self, snapshot):
        help_output = execute_in_terminal(
            ["copernicusmarine", "subset", "--help"]
        )

        assert str(help_output).split("\\n") == snapshot

    def test_help_from_login_is_as_expected(self, snapshot):
        help_output = execute_in_terminal(
            ["copernicusmarine", "login", "--help"]
        )

        assert str(help_output).split("\\n") == snapshot

    def test_help_from_copernicusmarin_is_as_expected(self, snapshot):
        help_output = execute_in_terminal(["copernicusmarine", "--help"])

        assert str(help_output).split("\\n") == snapshot
