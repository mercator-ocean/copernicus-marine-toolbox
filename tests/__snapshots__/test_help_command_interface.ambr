# serializer version: 1
# name: TestHelpCommandLineInterface.test_help_from_copernicusmarin_is_as_expected
  list([
    "CompletedProcess(args=['copernicusmarine', '--help'], returncode=0, stdout=b'Usage: copernicusmarine [OPTIONS] COMMAND [ARGS]...",
    '',
    'Options:',
    '  -V, --version  Show the version and exit.',
    '  -h, --help     Show this message and exit.',
    '',
    'Commands:',
    '  describe  Print Copernicus Marine catalogue as JSON.',
    '  get       Download originally produced data files.',
    '  login     Create a configuration file with your Copernicus Marine',
    '            credentials.',
    '  subset    Download subsets of datasets as NetCDF files or Zarr stores.',
    "', stderr=b'')",
  ])
# ---
# name: TestHelpCommandLineInterface.test_help_from_describe_is_as_expected
  list([
    "CompletedProcess(args=['copernicusmarine', 'describe', '--help'], returncode=0, stdout=b'Usage: copernicusmarine describe [OPTIONS]",
    '',
    '  Print Copernicus Marine catalogue as JSON.',
    '',
    '  The default display contains information on the products, and more data can',
    '  be displayed using the ``--include-<argument>`` flags.',
    '',
    '  The ``--contains`` option allows the user to specify one or several strings',
    '  to filter through the catalogue display. The search is performed recursively',
    '  on all attributes of the catalog, and the tokens only need to be contained',
    '  in one of the attributes (i.e., not an exact match).',
    '',
    'Options:',
    '  --include-description           Include product description in output.',
    '  --include-datasets              Include product dataset details in output.',
    '  --include-keywords              Include product keyword details in output.',
    '  --include-versions              Include dataset versions in output. By',
    '                                  default, shows only the default version.',
    '  -a, --include-all               Include all the possible data in output:',
    '                                  description, datasets, keywords, and',
    '                                  versions.',
    '  -c, --contains TEXT             Filter catalogue output. Returns products',
    '                                  with attributes matching a string token.',
    '  --max-concurrent-requests INTEGER',
    '                                  Maximum number of concurrent requests.',
    '                                  Default 15. The describe command uses a',
    '                                  thread pool executor to manage concurrent',
    '                                  requests.',
    '  --disable-progress-bar          Flag to hide progress bar.',
    '  --log-level [DEBUG|INFO|WARN|ERROR|CRITICAL|QUIET]',
    '                                  Set the details printed to console by the',
    '                                  command (based on standard logging library).',
    '  -h, --help                      Show this message and exit.',
    '',
    '  .. code-block:: bash',
    '',
    '      copernicusmarine describe --contains METOFFICE-GLO-SST-L4-NRT-OBS-SST-V2',
    '      --include-datasets',
    '',
    '  .. code-block:: bash',
    '',
    '      copernicusmarine describe -c METOFFICE-GLO-SST-L4-NRT-OBS-SST-V2',
    "', stderr=b'')",
  ])
# ---
# name: TestHelpCommandLineInterface.test_help_from_get_is_as_expected
  list([
    "CompletedProcess(args=['copernicusmarine', 'get', '--help'], returncode=0, stdout=b'Usage: copernicusmarine get [OPTIONS]",
    '',
    '  Download originally produced data files.',
    '',
    '  The ``--dataset-id`` is required (can be found via the "describe" command).',
    '  The function fetches the files recursively if a folder path is passed as a',
    '  URL. When provided a dataset ID, all the files in the corresponding folder',
    '  will be downloaded if none of the ``--filter`` or ``--regex`` options is',
    '  specified.',
    '',
    'Options:',
    '  -i, --dataset-id TEXT           The datasetID.',
    '  --dataset-version TEXT          Force the selection of a specific dataset',
    '                                  version.',
    '  --dataset-part TEXT             Force the selection of a specific dataset',
    '                                  part.',
    '  --username TEXT                 If not set, search for environment variable',
    '                                  COPERNICUSMARINE_SERVICE_USERNAME, or else',
    '                                  look for configuration files, or else ask',
    '                                  for user input.',
    '  --password TEXT                 If not set, search for environment variable',
    '                                  COPERNICUSMARINE_SERVICE_PASSWORD, or else',
    '                                  look for configuration files, or else ask',
    '                                  for user input.',
    '  -nd, --no-directories           Option to not recreate folder hierarchy in',
    '                                  ouput directory.                 \\x08 NOTE:',
    '                                  This argument is mutually exclusive with',
    '                                  arguments: [sync].',
    '  --show-outputnames              Option to display the names of the output',
    '                                  files before download.',
    '  -o, --output-directory PATH     The destination directory for the downloaded',
    '                                  files. Default is the current directory.',
    '  --credentials-file PATH         Path to a credentials file if not in its',
    '                                  default directory. Accepts',
    '                                  .copernicusmarine-credentials / .netrc or',
    '                                  _netrc / motuclient-python.ini files.',
    '  --force-download                Flag to skip confirmation before download.',
    '  --overwrite-output-data, --overwrite',
    '                                  If specified and if the file already exists',
    '                                  on destination, then it will be overwritten',
    '                                  instead of creating new one with unique',
    '                                  index.',
    '  --create-template               Option to create a file get_template.json in',
    '                                  your current directory containing CLI',
    '                                  arguments. If specified, no other action',
    '                                  will be performed.',
    '  --request-file PATH             Option to pass a file containing CLI',
    '                                  arguments. The file MUST follow the',
    "                                  structure of dataclass \\'GetRequest\\'. For",
    '                                  more information please refer to the README.',
    '  --filter, --filter-with-globbing-pattern TEXT',
    '                                  A pattern that must match the absolute paths',
    '                                  of the files to download.',
    '  --regex, --filter-with-regular-expression TEXT',
    '                                  The regular expression that must match the',
    '                                  absolute paths of the files to download.',
    '  --file-list PATH                Path to a .txt file containing a list of',
    '                                  file paths, line by line, that will be',
    '                                  downloaded directly. These files must be',
    '                                  from the specified dataset using the',
    '                                  --dataset-id. If no files can be found, the',
    '                                  Toolbox will list all files on the remote',
    '                                  server and attempt to find a match.',
    '  --create-file-list TEXT         Option to only create a file containing the',
    '                                  names of the targeted files instead of',
    '                                  downloading them. It writes the file in the',
    '                                  directory specified with the --output-',
    '                                  directory option (default to current',
    '                                  directory). The file name specified should',
    "                                  end with \\'.txt\\' or \\'.csv\\' If specified, no",
    '                                  other action will be performed. Please find',
    '                                  more information in the README.',
    '  --sync                          Option to synchronize the local directory',
    '                                  with the remote directory. See the',
    '                                  documentation for more details.',
    '                                  \\x08 NOTE: This argument is mutually exclusive',
    '                                  with arguments: [no-directories].',
    '  --sync-delete                   Option to delete local files that are not',
    '                                  present on the remote server while applying',
    '                                  sync.                 \\x08 NOTE: This argument',
    '                                  is mutually exclusive with arguments: [no-',
    '                                  directories].',
    '  --index-parts                   Option to get the index files of an INSITU',
    '                                  dataset. Temporary option.',
    '  --dry-run                       Runs query without downloading data.',
    '  --max-concurrent-requests INTEGER',
    '                                  Maximum number of concurrent requests.',
    '                                  Default 15. The get command uses a thread',
    '                                  pool executor to manage concurrent requests.',
    '  --disable-progress-bar          Flag to hide progress bar.',
    '  --log-level [DEBUG|INFO|WARN|ERROR|CRITICAL|QUIET]',
    '                                  Set the details printed to console by the',
    '                                  command (based on standard logging library).',
    '  -h, --help                      Show this message and exit.',
    '',
    '  .. code-block:: bash',
    '',
    '      # Example: Download all files from a dataset     copernicusmarine get -i',
    '      cmems_mod_nws_bgc-pft_myint_7km-3D-diato_P1M-m',
    "', stderr=b'')",
  ])
# ---
# name: TestHelpCommandLineInterface.test_help_from_login_is_as_expected
  list([
    "CompletedProcess(args=['copernicusmarine', 'login', '--help'], returncode=0, stdout=b'Usage: copernicusmarine login [OPTIONS]",
    '',
    '  Create a configuration file with your Copernicus Marine credentials.',
    '',
    '  Create a configuration file under the ``$HOME/.copernicusmarine`` directory',
    '  (overwritable with the ``--credentials-file`` option).',
    '',
    'Options:',
    '  --username TEXT                 If not set, search for environment variable',
    '                                  COPERNICUSMARINE_SERVICE_USERNAME, or else',
    '                                  ask for user input.',
    '  --password TEXT                 If not set, search for environment variable',
    '                                  COPERNICUSMARINE_SERVICE_PASSWORD, or else',
    '                                  ask for user input.',
    '  --configuration-file-directory PATH',
    '                                  Path to the directory where the',
    '                                  configuration file is stored.',
    '  -overwrite, --overwrite-configuration-file',
    '                                  Flag to skip confirmation before overwriting',
    '                                  configuration file.',
    '  --skip-if-user-logged-in        Flag to skip the logging process if the user',
    '                                  is already logged in.',
    '  --log-level [DEBUG|INFO|WARN|ERROR|CRITICAL|QUIET]',
    '                                  Set the details printed to console by the',
    '                                  command (based on standard logging library).',
    '  -h, --help                      Show this message and exit.',
    '',
    '  .. code-block:: bash',
    '',
    '      COPERNICUSMARINE_SERVICE_USERNAME=<USERNAME>',
    '      COPERNICUSMARINE_SERVICE_PASSWORD=<PASSWORD> copernicusmarine login',
    '',
    '  .. code-block:: bash',
    '',
    '      copernicusmarine login --username <USERNAME> --password <PASSWORD>',
    '',
    '  .. code-block:: bash',
    '',
    '      copernicusmarine login     > Username: [USER-INPUT]     > Password:',
    '      [USER-INPUT]',
    "', stderr=b'')",
  ])
# ---
# name: TestHelpCommandLineInterface.test_help_from_subset_is_as_expected
  list([
    "CompletedProcess(args=['copernicusmarine', 'subset', '--help'], returncode=0, stdout=b'Usage: copernicusmarine subset [OPTIONS]",
    '',
    '  Download subsets of datasets as NetCDF files or Zarr stores.',
    '',
    '  The ``--dataset-id`` is required (can be found via the "describe" command).',
    '  The argument values passed individually through the CLI take precedence over',
    '  the values from the ``--motu-api-request`` option, which takes precedence',
    '  over the ones from the ``--request-file`` option.',
    '',
    'Options:',
    '  -i, --dataset-id TEXT           The datasetID.',
    '  --dataset-version TEXT          Force the selection of a specific dataset',
    '                                  version.',
    '  --dataset-part TEXT             Force the selection of a specific dataset',
    '                                  part.',
    '  --username TEXT                 If not set, search for environment variable',
    '                                  COPERNICUSMARINE_SERVICE_USERNAME, or else',
    '                                  look for configuration files, or else ask',
    '                                  for user input.',
    '  --password TEXT                 If not set, search for environment variable',
    '                                  COPERNICUSMARINE_SERVICE_PASSWORD, or else',
    '                                  look for configuration files, or else ask',
    '                                  for user input.',
    '  -v, --variable TEXT             Specify dataset variable. Can be used',
    '                                  multiple times.',
    '  -x, --minimum-longitude FLOAT   Minimum longitude for the subset. The value',
    '                                  will be reduced to the interval [-180; 360[.',
    '  -X, --maximum-longitude FLOAT   Maximum longitude for the subset. The value',
    '                                  will be reduced to the interval [-180; 360[.',
    '  -y, --minimum-latitude FLOAT RANGE',
    '                                  Minimum latitude for the subset. Requires a',
    '                                  float within this range:  [-90<=x<=90]',
    '  -Y, --maximum-latitude FLOAT RANGE',
    '                                  Maximum latitude for the subset. Requires a',
    '                                  float within this range:  [-90<=x<=90]',
    '  -z, --minimum-depth FLOAT RANGE',
    '                                  Minimum depth for the subset. Requires a',
    '                                  float within this range:  [x>=0]',
    '  -Z, --maximum-depth FLOAT RANGE',
    '                                  Maximum depth for the subset. Requires a',
    '                                  float within this range:  [x>=0]',
    '  -V, --vertical-dimension-output [depth|elevation]',
    '                                  Consolidate the vertical dimension (the',
    '                                  z-axis) as requested: `depth` with',
    '                                  descending positive values. `elevation` with',
    '                                  ascending positive values. Default is',
    '                                  `depth`.',
    '  -t, --start-datetime TEXT       The start datetime of the temporal subset.',
    '                                  Caution: encapsulate date with " " to ensure',
    '                                  valid expression for format "%Y-%m-%d',
    '                                  %H:%M:%S". Supports common format parsed by',
    '                                  pendulum. See',
    '                                  https://pendulum.eustace.io/docs/#parsing',
    '  -T, --end-datetime TEXT         The end datetime of the temporal subset.',
    '                                  Caution: encapsulate date with " " to ensure',
    '                                  valid expression for format "%Y-%m-%d',
    '                                  %H:%M:%S". Supports common format parsed by',
    '                                  pendulum. See',
    '                                  https://pendulum.eustace.io/docs/#parsing',
    '  --coordinates-selection-method [inside|strict-inside|nearest|outside]',
    '                                  The method in which the coordinates will be',
    "                                  retrieved. If \\'inside\\', the retrieved",
    '                                  selection will be inside the requested',
    "                                  interval. If \\'strict-inside\\', the retrieved",
    '                                  selection will be inside the requested',
    '                                  interval and an error will raise if there',
    "                                  doesn\\'t exist the values. If \\'nearest\\', the",
    '                                  returned interval extremes will be the',
    '                                  closest to what has been asked for. A',
    '                                  warning will be displayed if outside of',
    "                                  bounds. If \\'outisde\\', the extremes will be",
    '                                  taken to contain all the requested interval.',
    '                                  A warning will also be displayed if the',
    '                                  subset is outside of the dataset bounds.',
    '  -o, --output-directory PATH     The destination folder for the downloaded',
    '                                  files. Default is the current directory.',
    '  --credentials-file PATH         Path to a credentials file if not in its',
    '                                  default directory. Accepts',
    '                                  .copernicusmarine-credentials / .netrc or',
    '                                  _netrc / motuclient-python.ini files.',
    '  -f, --output-filename TEXT      Concatenate the downloaded data in the given',
    '                                  file name (under the output directory).',
    '  --file-format [netcdf|zarr]     Format of the downloaded dataset. Default to',
    '                                  NetCDF (.nc).',
    '  --force-download                Flag to skip confirmation before download.',
    '  --overwrite-output-data, --overwrite',
    '                                  If specified and if the file already exists',
    '                                  on destination, then it will be overwritten',
    '                                  instead of creating new one with unique',
    '                                  index.',
    '  -s, --service TEXT              Force download through one of the available',
    '                                  services using the service name among',
    "                                  [\\'arco-geo-series\\', \\'arco-time-series\\',",
    "                                  \\'omi-arco\\', \\'static-arco\\'] or its short name",
    "                                  among [\\'geoseries\\', \\'timeseries\\', \\'omi-",
    "                                  arco\\', \\'static-arco\\'].",
    '  --create-template               Option to create a file subset_template.json',
    '                                  in your current directory containing CLI',
    '                                  arguments. If specified, no other action',
    '                                  will be performed.',
    '  --request-file PATH             Option to pass a file containing CLI',
    '                                  arguments. The file MUST follow the',
    "                                  structure of dataclass \\'SubsetRequest\\'. For",
    '                                  more information please refer to the README.',
    '  --motu-api-request TEXT         Option to pass a complete MOTU API request',
    '                                  as a string. Caution, user has to replace',
    '                                  double quotes " with single quotes \\\' in the',
    '                                  request.',
    '  --dry-run                       Runs query without downloading data.',
    '  --disable-progress-bar          Flag to hide progress bar.',
    '  --log-level [DEBUG|INFO|WARN|ERROR|CRITICAL|QUIET]',
    '                                  Set the details printed to console by the',
    '                                  command (based on standard logging library).',
    '  --netcdf-compression-enabled    Enable compression level 1 to the NetCDF',
    '                                  output file. Use --netcdf-compression-level',
    '                                  option to customize the compression level',
    '  --netcdf-compression-level INTEGER RANGE',
    '                                  Specify a compression level to apply on the',
    '                                  NetCDF output file. A value of 0 means no',
    '                                  compression, and 9 is the highest level of',
    '                                  compression available  [0<=x<=9]',
    '  --netcdf3-compatible            Enable downloading the dataset in a netCDF 3',
    '                                  compatible format.',
    '  -h, --help                      Show this message and exit.',
    '',
    '  .. code-block:: bash',
    '',
    '      copernicusmarine subset --dataset-id',
    '      cmems_mod_ibi_phy_my_0.083deg-3D_P1D-m --variable thetao --variable so',
    '      --start-datetime 2021-01-01 --end-datetime 2021-01-03 --minimum-',
    '      longitude 0.0 --maximum-longitude 0.1 --minimum-latitude 28.0 --maximum-',
    '      latitude 28.1 --minimum-depth 1 --maximum-depth 2',
    '',
    '  Equivalent to:',
    '',
    '  .. code-block:: bash',
    '',
    '      copernicusmarine subset -i cmems_mod_ibi_phy_my_0.083deg-3D_P1D-m -v',
    '      thetao -v so -t 2021-01-01 -T 2021-01-03 -x 0.0 -X 0.1 -y 28.0 -Y 28.1',
    '      -z 1 -Z 2',
    "', stderr=b'')",
  ])
# ---
